### ISSUE TITLE
Unable to set phase2alg in config file

### ISSUE THREAD

**Issue by lupien** [libreswan/libreswan#126]
I am using libreswan version 3.21 (fedora libreswan-3.21-1.fc26.x86_64).
I am using ipsec between this linux machine and a windows 7 machine using IKEv1 secret authentication and type transport (tunnelling tcp ports). The setup has been working for the last 4 years.

The windows default uses modp1024(dh2), so since my update to 3.21 (I used to have 3.18) I had to force the IKE encryption by using
  `ike=aes_cbc128-sha1;modp1024`
And with that change, it works fine.

So, I also tried to force the ESP encryption but I never got that to work.
Without specifying it (using the defaults), it works correctly and I get in the logs:
```
#219: intiating Quick Mode PSK+ENCRYPT+UP+IKEV1_ALLOW+IKEV2_ALLOW+SAREF_TRACK+IKE_FRAG_ALLOW+ESN_NO to replace #210 {using isakmp#179 msgid:b62d7324 proposal=defaults pfsgroup=no-pfs}
#219: transition from state STATE_QUICK_I1 to state STATE_QUICK_I2
#219: STATE_QUICK_I2: sent QI2, IPsec SA established transport mode {ESP=>0x41fbd7d2 <0xa42393a8 xfrm=AES_128-HMAC_SHA1 NATOA=none NATD=none DPD=passive}
```
When I try to specify it like
 `phase2alg=aes_ccm128-hmac_sha1`
or
 `phase2alg=aes-sha1`
or
 `phase2alg=aes128-sha1`

I always get the same problem:
```
#15: initiating Quick Mode PSK+ENCRYPT+UP+IKEV1_ALLOW+IKEV2_ALLOW+SAREF_TRACK+IKE_FRAG_ALLOW+ESN_NO {using isakmp#4 msgid:42210e2b proposal=NULL(11)_000-SHA1(2) pfsgroup=no-pfs}
#16: transition from state STATE_QUICK_I1 to state STATE_QUICK_I2
#16: STATE_QUICK_I2: sent QI2, IPsec SA established transport mode {ESP=>0x3e967b2a <0xe8b9b429 xfrm=NULL_0-HMAC_SHA1 NATOA=none NATD=none DPD=passive}
```
Observe that the proposal and established crypto are NULL instead of aes.
And this does not make a working ipsec connection for me.

Am I writing the phase2alg entry incorrectly or is there a problem parsing it?
I can't find any warning that something is not valid in the log file (beside those NULL).

Note that `ip xfrm state` shows
  `enc ecb(cipher_null)`
in that case instead of the
  `enc cbc(aes) 0xsomehexnumber....`
when it works.

**Comment by letoams** [comment]
On Wed, 13 Sep 2017, Christian Lupien wrote:

> I am using libreswan version 3.21 (fedora libreswan-3.21-1.fc26.x86_64).

> So, I also tried to force the ESP encryption but I never got that to work.
> Without specifying it (using the defaults), it works correctly and I get in the logs:
> 
> #219: intiating Quick Mode PSK+ENCRYPT+UP+IKEV1_ALLOW+IKEV2_ALLOW+SAREF_TRACK+IKE_FRAG_ALLOW+ESN_NO to replace #210 {using isakmp#179 msgid:b62d7324 proposal=defaults pf
> sgroup=no-pfs}
> #219: transition from state STATE_QUICK_I1 to state STATE_QUICK_I2
> #219: STATE_QUICK_I2: sent QI2, IPsec SA established transport mode {ESP=>0x41fbd7d2 <0xa42393a8 xfrm=AES_128-HMAC_SHA1 NATOA=none NATD=none DPD=passive}
> 
> When I try to specify it like
> phase2alg=aes_ccm128-hmac_sha1

aes_ccm is wrong here, as it requires a null hash algo.

> or
> phase2alg=aes-sha1
> or
> phase2alg=aes128-sha1
> 
> I always get the same problem:

The last two should work.

> #15: initiating Quick Mode PSK+ENCRYPT+UP+IKEV1_ALLOW+IKEV2_ALLOW+SAREF_TRACK+IKE_FRAG_ALLOW+ESN_NO {using isakmp#4 msgid:42210e2b proposal=NULL(11)_000-SHA1(2) pfsgroup
> =no-pfs}
> #16: transition from state STATE_QUICK_I1 to state STATE_QUICK_I2
> #16: STATE_QUICK_I2: sent QI2, IPsec SA established transport mode {ESP=>0x3e967b2a <0xe8b9b429 xfrm=NULL_0-HMAC_SHA1 NATOA=none NATD=none DPD=passive}
> 
> Observe that the proposal and established crypto are NULL instead of aes.
> And this does not make a working ipsec connection for me.

> Note that ip xfrm state shows
> enc ecb(cipher_null)
> in that case instead of the
> enc cbc(aes) 0xsomehexnumber....
> when it works.

That's really odd. I took the 3.21-1 fc26 and recompiled it on my older
fedora and ran a testrun and it gave me:

004 "westnet-eastnet" #2: STATE_QUICK_I2: sent QI2, IPsec SA established tunnel mode {ESP=>0xESPESP <0xESPESP xfrm=AES_128-HMAC_SHA1 NATOA=none NATD=none DPD=passive}

and ip xfrm pol gave me:

        auth-trunc hmac(sha1) 0xHASHKEY 96
        enc cbc(aes) 0xENCKEY

I don't have a good explanation of what you are seeing. Our test cases
would have also screamed murder if that was happening.

Can you verify this is really still happening for you and not a weird
glitch? Are you sure you have only 1 pluto installed via rpm and not
another one in /usr/local/libexec/ipsec/pluto ?

Paul

**Comment by lupien** [comment]
I found the source of the problem.
I only have one copy of the software installed.
However my config files were causing the problem,

But it is reacting in a weird way (maybe by design?)
If I make the attached file the /etc/ipsec.conf file, then the encoding is correct.
However if a make the attached file /etc/ipsec2.conf and I create an ipsec.conf file containing only
`include /etc/ipsec2.conf`
then it fails like I described above (that is similar to the configuration I was using).
Note that the re7 entry was a leftover from previous tests. I am not using it.

So to reproduce the bad encoding I was seeing, I needed to include a file, with a bad phase2alg entry in an ignored section followed by a %default section that defines a good phase2alg.

My fix was to comment that bad section.

[ipsec.txt](https://github.com/libreswan/libreswan/files/1327851/ipsec.txt)
